### PR TITLE
EDGEPOD-397: Handling of HO Cancel, HO Failure

### PR DESCRIPTION
**Pull request by asreek**

### CODE REVIEW

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

File: src/mme-app/actionHandlers/s1HandoverActionHandlers.cpp [hunk 1]
```diff
@@ -434,3 +447,213 @@ ActStatus ActionHandlers::is_tau_required(ControlBlock& cb)
...
+    }
+   
+    struct s1relcmd_info s1relcmd;
+    s1relcmd.msg_type = s1_release_command;
+    s1relcmd.ue_idx = ue_ctxt->getContextID();
+    s1relcmd.enb_s1ap_ue_id = ho_ctxt->getTargetS1apEnbUeId();
+    s1relcmd.cause = ho_ctxt->getS1HoCause().s1apCause_m;
+    s1relcmd.enb_fd = ho_ctxt->getTargetEnbContextId();
+
+    // Fire and forget s1 release to target enb
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/s1HandoverActionHandlers.cpp:570]: Is this per spec ? If not open a enhancement defect to handle this later.

File: src/mmeGrpcProtos/mmeGrpc.proto [hunk 2]
```diff
@@ -114,7 +114,11 @@ message ProcedureStatsRespBuf
 	uint32 num_of_ho_request_ack_received = 55;
 	uint32 num_of_ho_command_to_src_enb_sent = 56;
 	uint32 num_of_ho_notify_received = 57;
-	uint32 num_of_ho_complete = 58;	
+	uint32 num_of_ho_complete = 58;
```
Comment (thakurajayL) [inline src/mmeGrpcProtos/mmeGrpc.proto:117]: wherever possible we should use uint32 or similar variable names...this really makes code portable. Other places counters are using int variable type which is not so good idea.

File: src/s1ap/handlers/handover_cancel.c [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            log_msg(LOG_INFO, "handover cancel S1AP_IE_CAUSE.\n");
+
+            memcpy(&ho_cancel.msg_data.handover_cancel_Q_msg_m.cause,
+                    &ho_cancel_ies.data[i].val.cause,
+                    sizeof(struct s1apCause));
+        }
+            break;
+        default:
+            log_msg(LOG_WARNING, "Unhandled IE");
```
Comment (thakurajayL) [inline src/s1ap/handlers/handover_cancel.c:65]: Good idea to put ie_type number in the default switch. Good practice to do it.

File: src/s1ap/handlers/s1ap_encoder.c [hunk 4]
```diff
@@ -1267,3 +1267,124 @@ int s1ap_mme_encode_handover_mme_status_transfer(
     *length = enc_ret;
     return enc_ret;
 }
+
+int s1ap_mme_encode_handover_prep_failure(
+  struct handover_preparation_failure_Q_msg *s1apPDU,
+  uint8_t **buffer,
+  uint32_t *length)
+{
+    S1AP_PDU_t pdu = { (S1AP_PDU_PR_NOTHING) };
```
Comment (thakurajayL) [inline src/s1ap/handlers/s1ap_encoder.c:1276]: we are doing memset below. In that case initializing pdu here may not be required.

**Pull request merged**